### DISCUSSION TITLE
Using Source Generators to create methods

### DISCUSSION THREAD

**Discussion by mrlacey** [CommunityToolkit/Maui.Markup#145] [General]
Following on from https://github.com/CommunityToolkit/Maui.Markup/issues/142#issuecomment-1311496665

What's the benefit of using Source Generators here?

Doesn't it make testing and documenting of the available functionality harder?

In that the amount of effort to write the code for the generator is approximately equivalent to writing the methods that would be generated directly, the fact that what will be generated will (almost?) always be the same it looks like work is being unnecessarily duplicated and offloaded to the consumer of the library when it could be done once.
It's the difference between writing and compiling the code once when the library is created, or generating and compiling that code every time every consuming app is built. That feels like a waste of compilation effort by duplicating the task (even though, admittedly very small) for every app that's built rather than just doing it once when the library is created.

The only benefit I see with the way that the current Source Generator implementation works with `ITextAlignment` means that it it will automatically add support for any new controls in the Microsoft.Maui.Controls assembly that also implement this interface.
However I'd expect (although I haven't yet checked) it would be possible to achieve the same things by building an extension based on the interface.

It _feels_ like the use of Source Generators is unnecessary and doesn't really add any real benefit. It just makes the codebase more complicated as there are some features and required markup functionality that can't be created this way.

**Comment by TheCodeTraveler** [comment]
> However I'd expect (although I haven't yet checked) it would be possible to achieve the same things by building an extension based on the interface

You've actually nailed the problem! All of the interfaces in .NET MAUI are read-only; the interface properties only have getters, no setters. 

So the only way to set a property based on the usage of a `Microsoft.Maui` interface and ensure we cover _every_ usage of the interface in both .NET MAUI and the user's custom controls is to use Source Generators to detect the interface and generate the extension methods.

**Reply by bijington** [reply]
It does sound like just writing the code will be the best approach. Especially as we need to document it too

**Reply by mrlacey** [reply]
Yes, for custom controls that implement an interface (such as `ITextAlignment`) and that you want to provide methods for, a SourceGenerator is a great solution.
My concern is using that for the controls in the Microsoft.Maui.Controls assembly. 

How much support there should be for custom and 3rd party controls doesn't seem to have been documented or discussed (apart from, minimally, at https://github.com/CommunityToolkit/Maui.Markup/discussions/86).

**Comment by pictos** [comment]
@mrlacey, actually using SGs (Incremental Source Generators, in this case) can be less code to maintain and more flexibility. So for reference, I'll take your PR #141, here all the methods just grab a value, set it to the property, and return the `object`, as in the snippet below:

```csharp
public static TVisualElement Width<TVisualElement>(this TVisualElement element, double widthRequest) where TVisualElement : VisualElement
{
	element.WidthRequest = widthRequest;
	return element;
}
```

So in this case we do it for all properties inside `VisualElement`.

From a general point of view, you can say that SG is a `reflection` for the compile time, you can access all members and info on an `object`, that said, with SGs here, we can create a `for loop` that will emit those methods for all `public` properties inside the `VisualElement`, and the method will set the property and return the `object`.

> What's the benefit of using Source Generators here?

So at the end of the day, we will have one code that will generate those methods and if in the future .NET MAUI add more properties, the SG will see that and generate the methods, so we don't need to open a PR adding more extensions methods. Of course, there are cases where writing down the methods is a best solution, for cases like this having a SG, IMHO is better.

> or generating and compiling that code every time every consuming app is built.

This isn't true, because we use `Incremental Source Generators`, so the code generation will occours just when something changed. So for controls that are inside .NET MAUI the generation will happens just once. For customControl the generation will happen just when the dev change something that requires a code generation, for example creating or editing a property. 

> It feels like the use of Source Generators is unnecessary and doesn't really add any real benefit. It just makes the codebase more complicated as there are some features and required markup functionality that can't be created this way.

Why do you've this feeling? TBH, writing a SG for this isn't that hard. If you never wrote one I'll be more than happy to guide you through the process (if we choose to re-implement your PR using SG)

**Reply by mrlacey** [reply]
My apologies, I thought you were shipping the generator, not using it only within the project

**Reply by pictos** [reply]
@mrlacey not sure if I follow you... The idea is to ship the SG with the package, we already have an SG that implements the `TextAlignmentExtensions`, as you can see [here](https://github.com/CommunityToolkit/Maui.Markup/tree/main/src/CommunityToolkit.Maui.Markup.SourceGenerators)

**Reply by mrlacey** [reply]
There are a few issues that I think are being jumbled together here. (Some by everyone, some maybe just by me.)

**Firstly, Using Source Generators for custom controls and 3rd party libraries**
This is what is done (in addition to other things--that I'll get to) by the current SG implementation for anything implementing the `ITextAlignment` interface.  
How, where, when, and why to extend classes beyond the core MAUI controls probably needs a separate, longer, documented discussion. I can see pros and cons to doing this but also scenarios where it might be useful (necessary?) for someone creating a control to be able to opt-out of having extension methods generated in this way.
Let me put a pin in this issue for now.


**Secondly Using Source Generators to create extension methods for classes inside the Microsoft.Maui.Controls assembly**

As the current SG implementation works, every consuming application will end up with something like the following: 

![partial screenshot showing the classes generated by the SG](https://user-images.githubusercontent.com/189547/201467291-9891f938-9a71-48c3-aba7-983cb3f0eebc.png)

These files will "always" be the same and must be compiled into each and every app when built.
Ok, so "always" is only true as long as the Microsoft.Maui.Controls library doesn't change. But how often is this really likely to happen?
If the library is really expected to change a lot, that raises a bunch of concerns for someone considering building with MAUI.  
If the intention is to support changes in MAUI happening much faster than this library can keep up with, that raises concerns for anyone considering using this library.

What using Source generators this way does is mean that rather than shipping a library with some compiled methods, a library is shipped that generates the same source that every consuming application must then compile.

Maybe it's because I'm old, but I've learned to be concerned about what source is compiled, when it's compiled, and the benefits of shipping a library rather than shipping source.
Similarly, when the output is always the same, there are benefits to doing something once, centrally, rather than having every consumer do the same thing. (In this case the "something" is the reflecting over assemblies to find the classes to generate source for.)
It might be easy to argue that modern machines are fast and compilation is quick. But is it still worth doing work unnecessarily?
It could be argued that (based on the above image) I'm only talking about 6 small files/classes. But if the use of SGs here is to become a standard within this project, there will quickly become a lot more. 
All this extra work (generation and compilation) isn't free. Yes, on its own, it's not enough to change the world. Still, by requiring all this extra processing and compilation in each consuming application, more work is being done overall, and this requires power that has a cost--both financially and in the use of resources to produce that power.

If my above concerns are factors that have been considered unimportant or an acceptable trade-off, then fine. I'd like to understand why. Or at least know that they were considered.  
The code in this project is seen by many as an example of how to do things, and I'm concerned about anyone setting a precedent of always generating the same code in a consuming project rather than shipping a compiled assembly.

Yes, source generators are great and are useful for many situations. Always generating the same code in every app isn't one.


Going back to the properties of `VisualElement,` yes, a Source Generator could be created to iterate over the settable properties and output simple methods for each of them. 
But why stop there? 
Surely it would be more beneficial (and little extra work) to iterate over every control that inherits from `Microsoft.Maui.Controls.View` (or similar--I'm open to suggestions) and create methods for every accessible settable property. This would allow the focus to be on additional benefits that other methods (or features or functionality) could provide.
I'm assuming that there was a reason this wasn't done when the current Source Generator was created. What was the reason?

My preference is that the library would include the output for controls from official MAUI assemblies and only run the generator for controls from elsewhere. This would avoid the unnecessary compilation of the same source code in every project but may raise other issues, such as how to opt-out of this functionality for specific assemblies, classes, or methods.  
It might even be appropriate to make such an SG a separate library so that the behavior of SG interaction with custom and 3rd party controls could be fully opt-in.

I don't want to rule out the use of code/source generation entirely. Even if the generators aren't used in the shipped library to generate code based on classes in the Microsoft.Maui.Controls assembly (and others?), they could still be used to produce the code that is compiled and shipped within the library. (Yes, I acknowledge this would require slightly modifying their use and the development workflow.)

Not relying on the use of generators for things that won't change also allows for tailored testing and documentation of the functionality that is shipped. There would be no need to rely on "generic" help messages, but specific, appropriate documentation can be provided. Any edge cases can also be thoroughly understood and tested for.

I have not made a final decision. I'm open to hearing about why the use of Source Generators in the way that it is currently done is preferable and the reasons for taking this approach.  As others have already (hopefully) spent more time than me thinking about this, I'd love to see some of that documented.  
The current lack of this documentation leads me to my final issue.


**Thirdly, There seems to be a lack of overall vision and planning** 

I admit this is at a different level from the above, but it underpins my interest here, and I hope serves as useful background.


My ultimate goal in looking at this library is to try and understand how using C# to define the UI can be preferable (even beneficial?) to using XAML.  
There are lots of factors here but having a complete, well-thought-out library to help with creating the C# code is part of that.  
At the moment, this library feels far from complete, and there isn't evidence of big-picture planning of the end goal and how to get there. Source Generators could well be part of acheiveing that end goal (whatever it is), but it's unclear how and why.

Having methods for all properties is only a small part of the potential benefit of using C# for markup.
The real value is in adding specific methods that add value by greatly simplifying the creation of the desired UI in a semantically clear way. 
Just chaining property setters doesn't add a lot. But doing more than this takes time and thought that can't be automated away.
Maybe having source generators create all the basic extension methods that directly map to properties is a good way to get the basics out of the way so it's possible to focus on the things that add real value. But I don't want to assume that's the best reason to use them, or why you've chosen to use them so far.





And yes, there are other possible options to consider, such as only generating extension methods for classes that are used, but I don't want to muddy the water here further. I'm sure those extra issues can be discussed in future, if necessary.

And to respond to the other direct comment above, no, the generator may not run on EVERY build, but it will be evaluated every time references are changed in the app. Also, the source would still get compiled every time the app is built.

My concern is that EVERY consuming app will do this.
While evaluation and generation of source might be quick, it's going to be done tens of thousands, or hundreds of thousands, or maybe millions of times (based on how popular the library is.) In that each of those generations is going to produce "exactly" the same thing, it seems like a massive, unnecessary duplication of effort.

**Reply by bijington** [reply]
I would like to apologise for the lack of detail that I have provided over the past few discussions, I have made statements about whether SGs will solve issues mainly as I asked the question of the team given I don't have a vast amount of experience with them and wanted to confirm whether we could find ways of keeping our implementation and maintenance effort as low as possible. I was under the hopeful impression that we could/would generate the source once when we build and then ship the resulting library. Perhaps this was a misunderstanding on my part.

I completely agree that SGs wouldn't add the real value where we add much more useful methods - one such one might be the `Text` method that allows you to do multiple things in a single method call.

I am also coming to the conclusion that while SGs might save us some effort on the writing of the methods (and yes we were hoping to expand upon many other types within .NET MAUI to expose extension methods such as View, Entry, etc. or at least find the most suitable base classes to cover), the resulting savings will likely not help us much by the time we still have to write up the documentation, etc.

In terms of overall vision and planning, we would love to cover the entire surface area of .NET MAUI controls in order to reach feature parity with XAML. I believe we have verbalised this but we likely don't have anything clearly defined on here - perhaps we should and even start adding in proposals for all of the items. In fact I like this concept as it opens up the door for contributions when things are already clearly defined. 

I have stated my reasoning on why we might consider SGs and I was hopeful we would get a discussion like we are having to help decide on whether they really are the answer for this. Thanks for this discussion.

**Reply by bijington** [reply]
I guess a lack of documentation is most likely down to the fact that we are only on the initial thought phase around using SGs, at least in terms of the Markup implementation for these methods. I chimed in quickly on your PR/proposal to try and save on any wasted effort if that was the chosen approach. 

I really value discussions like this to help us decide the best approach.

**Reply by idexus** [reply]
@bijington  

> I was under the hopeful impression that we could/would generate the source once when we build and then ship the resulting library. Perhaps this was a misunderstanding on my part. 

The code is generated once and then it can be sent together with the library, you can of course attach the generator to the library but then it is executed only on user types. This is what I did in [my library](https://github.com/idexus/Sharp.UI) to cover all properties and event handlers for all classes from `Microsoft.Maui.Controls` down to the `BindableObject' class

Documentation can also be generated with a little effort based on documentation from the MAUI project

**Reply by pictos** [reply]
@idexus, I saw that in your lib you use the `ISourceGenerator` instead of `IIncrementalSourceGenerator`, so your classes aren't generated just once. I would say it will generate the types every time something in the project changes. Anyway I recommend you to update your lib to use the new `IIncrementalSourceGenerator`.

**Reply by idexus** [reply]
@pictos Thanks, I'll have to do that over time as I also generate code based on user defined classes.

What I meant was that for already compiled libraries, when compiling a project using them, the generator for them is not restarted. I tested it.

**Reply by pictos** [reply]
@idexus do you think we can continue the discussion here or should I create a new one?

**Reply by idexus** [reply]
@pictos we can stay here or create a new discussion about broader use of SG in the context of creating a new foundation to cover all properties and create inline builders for them to create property bindings, theme dependency, platform etc. I will also be opening PR [https://github.com/dotnet/maui/issues/12678](https://github.com/dotnet/maui/issues/12678) for adding `IEnumerable` interface in maui project for classes with `[ContentProperty]` attribute, so we can do the rest by extension methods :)

**Comment by CliffAgius** [comment]
Closed as per the Community Stand-up discussion here: https://youtu.be/N9wMcBP4jtg?t=2889
